### PR TITLE
feat(mail): send mail with nodemailer, validate uuid and assing new passowrd to user [COM-43]

### PR DESCRIPTION
**Pull request by IsaacGzOv**
- Send link to the mail with the uuid
-Get password & uuid, validate uuid and save new password

[COM-43]

[COM-43]: https://comunidadesbootcamp.atlassian.net/browse/COM-43?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bd.sql [hunk 1]
```diff
@@ -19,6 +19,12 @@ create table users(
 	aliado VARCHAR(20)
 );
 
+create table mail(
+	id INT NOT NULL,
+	uuid VARCHAR(100),
+	FOREIGN KEY (id) REFERENCES users (id)
+);
+
```
Comment (gonzajo) [inline bd.sql]: change the name to something more explicit, like a `token`

File: sendMail.js [hunk 2]
```diff
@@ -3,7 +3,7 @@
 const nodemailer = require('nodemailer');
 require('dotenv').config();
 
-const from = 'isaac.gz.ov@gmail.com';
+const from = process.env.MAIL;
 const to = 'nevekdlh@gmail.com';
```
Comment (gonzajo) [inline sendMail.js]:
```suggestion
const from = process.env.MAIL_FROM;
const to = process.env.MAIL_TO';
```

File: src/controllers/user-controller.js [hunk 3]
```diff
@@ -18,6 +19,11 @@ const user = joi.object({
   aliado: joi.string().max(20).required(),
 });
 
+const newPass = joi.object({
+  uuid: joi.string().max(40).required(),
+  password: joi.string().required(),
+});
```
Comment (gonzajo) [inline src/controllers/user-controller.js]:
```suggestion
const newPass = joi.object({
  uuid: joi.string().guid({ version: 'uuidv4' }).required(),
  password: joi.string().required(),
});
```

also what is the expected validations for the password? length, etc...

File: src/repositories/mail-repository.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+const connect = require('../connection/dbconnection');
+
+const createNewUuid = async (id, uuid) => {
```
Comment (gonzajo) [inline src/repositories/mail-repository.js]:
```suggestion
const createToken = async (id, uuid) => {
```

File: src/repositories/mail-repository.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    return connection.query('INSERT INTO mail (id, uuid) VALUES (?,?)', [
+      id,
+      uuid,
+    ]);
+  } catch (error) {
+    throw { status: 500, message: error };
+  }
+};
+
+const findUuid = async (uuid) => {
```
Comment (gonzajo) [inline src/repositories/mail-repository.js]:
```suggestion
const findToken = async (uuid) => {
```

File: src/services/mail-service.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+const userReporsitory = require('../repositories/user-repository');
+const mailRepository = require('../repositories/mail-repository');
+
+// const API_KEY = process.env.SG_API_KEY;
+const fMail = process.env.MAIL;
+const url = process.env.URL;
+const text = '/recuperar-contrseña?token=';
+
+const sendMail = async (mail) => {
```
Comment (gonzajo) [inline src/services/mail-service.js]: update the name to something more meaningful to what is the method doing...

File: src/services/user-service.js [hunk 7]
```diff
@@ -40,10 +41,20 @@ const deleteUser = async (userId) => {
   }
 };
 
+const setNewPassword = async ({ uuid, password }) => {
+  try {
+    const { id } = await mailRepository.findUuid(uuid);
+    return await userRepository.setPassword(id, password);
```
Comment (gonzajo) [inline src/services/user-service.js]:
```suggestion
    return userRepository.setPassword(id, password);
```

File: src/routes/mail-routes.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+const express = require('express');
+const mailController = require('../controllers/mail-controller');
+
+const router = express.Router();
+
+router.get('/mail/sendMail/:mail', mailController.sendMail);
```
Comment (gonzajo) [inline src/routes/mail-routes.js]: update the route name, `user/recover-password` something more meaningful

File: src/controllers/user-controller.js [hunk 9]
```diff
@@ -120,10 +126,33 @@ const deleteUser = async (req, res) => {
...
+  if (result.error) {
+    res.status(400).send({
+      status: 'FAILED',
+      data: { error: result.error.details },
+    });
+  } else {
+    try {
+      await userService.setNewPassword(result.value);
+      res
+        .status(202)
```
Comment (gonzajo) [inline src/controllers/user-controller.js]: why 202?

File: src/controllers/user-controller.js [hunk 10]
```diff
@@ -120,10 +126,33 @@ const deleteUser = async (req, res) => {
...
+        .send({ status: 'OK', message: 'Change user password success' });
+    } catch (error) {
+      res.status(404).send({
```
Comment (gonzajo) [inline src/controllers/user-controller.js:143]: why 404?

**Review by gonzajo — COMMENTED** [review]
`src/repositories/mail-repository.js` change mail to `token...`

**Review by gonzajo — CHANGES_REQUESTED** [review]
.

File: src/controllers/user-controller.js (hunk 9 above)
Comment (IsaacGzOv) [reply src/controllers/user-controller.js]: im goin to change the code for 200, i was thinking that the code 202 was the correct one

File: src/controllers/user-controller.js (hunk 10 above)
Comment (IsaacGzOv) [reply src/controllers/user-controller.js:143]: Here because it didn't found the uuid

File: README.md [hunk 11]
```diff
@@ -3,12 +3,16 @@
...
+DATABASE =
 DATABASE_URL =
-SECRET_KEY = 
-PASS_MAIL = 
+SECRET_KEY =
+PASS_MAIL =
+SECRET_CRYPTR =
+SG_API_KEY =
+MAIL = 'startupchihuahua0@gmail.com'
+URL = 'https://startupchihuahua.onrender.com'
```
Comment (gonzajo) [inline README.md]: for the local env this should be pointing to localhost:3000 or the actual port where is react running on local
on the render side the env variable should be this one `https://startupchihuahua.onrender.com`

**Review by gonzajo — CHANGES_REQUESTED** [review]

File: bd.sql [hunk 12]
```diff
@@ -19,6 +19,12 @@ create table users(
 	aliado VARCHAR(20)
 );
 
+create table token(
+	id INT NOT NULL,
```
Comment (gonzajo) [inline bd.sql]: this should be the primary key right?

File: src/routes/token-routes.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+const express = require('express');
+const tokenController = require('../controllers/token-controller');
+
+const router = express.Router();
+
+router.get('/user/recover-password/:mail', tokenController.sendMail);
```
Comment (gonzajo) [inline src/routes/token-routes.js:6]: why `:mail`?

File: src/services/token-service.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+// const API_KEY = process.env.SG_API_KEY;
+const fMail = process.env.MAIL;
+const url = process.env.URL;
+const text = '/recuperar-contraseña?token=';
+
+const sendTokentoMail = async (mail) => {
+  const uuid = uuidv4();
+  const link = `${url}${text}${uuid}`;
+  console.log(link);
```
Comment (gonzajo) [inline src/services/token-service.js]: remove this

**Review by gonzajo — CHANGES_REQUESTED** [review]

File: src/routes/token-routes.js (hunk 13 above)
Comment (IsaacGzOv) [reply src/routes/token-routes.js:6]: here i send the email with the uuid

**Review by gonzajo — APPROVED** [review]

**Review by Isaac-G-O — APPROVED** [review]

**Pull request merged**